File: music_player/player.js
const express = require('express');
const { Client, GatewayIntentBits } = require('discord.js');
const {
    joinVoiceChannel,
    createAudioPlayer,
    createAudioResource,
    AudioPlayerStatus,
    VoiceConnectionStatus
} = require('@discordjs/voice');
const play = require('play-dl');

const client = new Client({
    intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildVoiceStates
    ]
});

// サーバーごとの接続情報を保存
const serverQueues = new Map();

client.on('ready', () => {
    console.log('Music Player Bot is online!');
});

const app = express();
app.use(express.json());
const port = 8080;

app.post('/play', async (req, res) => {
    const { guildId, channelId, query, userId } = req.body;
    if (!guildId || !channelId || !query || !userId) {
        return res.status(400).send({ error: 'リクエスト情報が不足しています。' });
    }

    try {
        const guild = await client.guilds.fetch(guildId);
        const member = await guild.members.fetch(userId);
        const textChannel = await guild.channels.fetch(channelId);

        if (!member.voice.channel) {
            return res.status(400).send({ error: 'まずボイスチャンネルに参加してください。' });
        }
        
        const queue = serverQueues.get(guildId);
        
        // 1. play-dlで動画を検索
        const searchResult = await play.search(query, { limit: 1 });
        if (searchResult.length === 0) {
            return res.status(404).send({ error: 'トラックが見つかりませんでした。' });
        }
        const track = searchResult[0];

        if (!queue) {
            // キューがない場合は、新しいキューを作成して接続
            const newQueue = {
                voiceChannel: member.voice.channel,
                textChannel: textChannel,
                connection: null,
                player: createAudioPlayer(),
                songs: [track]
            };

            serverQueues.set(guildId, newQueue);

            try {
                newQueue.connection = joinVoiceChannel({
                    channelId: member.voice.channel.id,
                    guildId: guild.id,
                    adapterCreator: guild.voiceAdapterCreator,
                });

                newQueue.connection.subscribe(newQueue.player);
                playNext(guildId);
                res.status(200).send(`✅ **${track.title}** の再生を開始します。`);

            } catch (err) {
                console.error(err);
                serverQueues.delete(guildId);
                return res.status(500).send({ error: 'ボイスチャンネルへの接続に失敗しました。' });
            }
        } else {
            // 既にキューがある場合は、曲を追加
            queue.songs.push(track);
            return res.status(200).send(`✅ **${track.title}** をキューに追加しました。`);
        }

    } catch (e) {
        console.error('Error in /play route:', e);
        return res.status(500).send({ error: `エラーが発生しました: ${e.message}` });
    }
});

async function playNext(guildId) {
    const queue = serverQueues.get(guildId);
    if (!queue) return;
    if (queue.songs.length === 0) {
        // キューが空になったらVCから切断
        if (queue.connection) {
            queue.connection.destroy();
        }
        serverQueues.delete(guildId);
        return;
    }

    const track = queue.songs.shift();

    try {
        const stream = await play.stream(track.url);
        const resource = createAudioResource(stream.stream, { inputType: stream.type });
        
        queue.player.play(resource);
        queue.textChannel.send(`🎵 再生中: **${track.title}**`);

        queue.player.once(AudioPlayerStatus.Idle, () => {
            playNext(guildId);
        });
    } catch (error) {
        console.error(`Error playing track: ${error}`);
        queue.textChannel.send(`❌ **${track.title}** の再生中にエラーが発生しました。`);
        playNext(guildId); // 次の曲へ
    }
}

app.post('/stop', (req, res) => {
    const guildId = req.body.guildId;
    const queue = serverQueues.get(guildId);

    if (queue) {
        queue.songs = []; // キューを空にする
        if (queue.player) queue.player.stop();
        if (queue.connection) queue.connection.destroy();
        serverQueues.delete(guildId);
        res.status(200).send({ message: '⏹️ 再生を停止しました。' });
    } else {
        res.status(400).send({ error: '現在再生中ではありません。' });
    }
});

app.post('/skip', (req, res) => {
    const guildId = req.body.guildId;
    const queue = serverQueues.get(guildId);
    if (!queue || queue.songs.length === 0) {
        return res.status(400).send({ error: 'スキップする曲がありません。' });
    }
    // プレイヤーを停止させると、'Idle'イベントが発火して次の曲が再生される
    queue.player.stop(); 
    res.status(200).send({ message: '⏭️ スキップしました。' });
});

client.login(process.env.DISCORD_BOT_TOKEN);
app.listen(port, () => {
    console.log(`Music player server listening at http://localhost:${port}`);
});

File: music_player/player.py
import flask
from flask import request, jsonify
import discord
from discord.ext import commands
import yt_dlp
import asyncio
import threading

# --- 初期設定 ---
intents = discord.Intents.default()
intents.message_content = True
bot = commands.Bot(command_prefix="!", intents=intents)
app = flask.Flask(__name__)

# yt-dlpの設定
YDL_OPTIONS = {'format': 'bestaudio/best', 'noplaylist': 'True', 'outtmpl': '%(extractor)s-%(id)s-%(title)s.%(ext)s'}
FFMPEG_OPTIONS = {'before_options': '-reconnect 1 -reconnect_streamed 1 -reconnect_delay_max 5', 'options': '-vn'}

# サーバーごとの音楽キュー
queues = {}

# --- Botの非同期処理 ---
@bot.event
async def on_ready():
    print(f'Music Player Logged in as {bot.user}')

async def play_next(guild_id):
    """次の曲を再生する"""
    if guild_id in queues and queues[guild_id]:
        # 次の曲の情報を取得
        url = queues[guild_id].pop(0)
        
        # 音声クライアントを取得
        voice_client = discord.utils.get(bot.voice_clients, guild_id=guild_id)
        if not voice_client:
            return

        # 音源を取得
        with yt_dlp.YoutubeDL(YDL_OPTIONS) as ydl:
            info = ydl.extract_info(url, download=False)
            audio_url = info['url']

        # 再生
        source = await discord.FFmpegOpusAudio.from_probe(audio_url, **FFMPEG_OPTIONS)
        voice_client.play(source, after=lambda e: asyncio.run_coroutine_threadsafe(play_next(guild_id), bot.loop).result())

# --- FlaskのAPIエンドポイント ---
@app.route('/play', methods=['POST'])
def play_song():
    data = request.get_json()
    guild_id = int(data['guild_id'])
    channel_id = int(data['channel_id'])
    query = data['query']

    async def _play():
        # ボイスチャンネルに接続
        guild = bot.get_guild(guild_id)
        if not guild:
            return jsonify({'error': 'Guild not found'}), 404
        
        voice_client = discord.utils.get(bot.voice_clients, guild=guild)
        if not voice_client:
            channel = guild.get_channel(channel_id)
            if not channel:
                return jsonify({'error': 'Channel not found'}), 404
            voice_client = await channel.connect()

        # キューに追加
        if guild_id not in queues:
            queues[guild_id] = []
        queues[guild_id].append(query)

        # 再生中でなければ再生開始
        if not voice_client.is_playing():
            await play_next(guild_id)

    future = asyncio.run_coroutine_threadsafe(_play(), bot.loop)
    future.result()
    return jsonify({'status': 'added to queue'})


@app.route('/skip', methods=['POST'])
def skip_song():
    data = request.get_json()
    guild_id = int(data['guild_id'])
    
    voice_client = discord.utils.get(bot.voice_clients, guild_id=guild_id)
    if voice_client and voice_client.is_playing():
        voice_client.stop()
        return jsonify({'status': 'skipped'})
    return jsonify({'error': 'Not playing anything'}), 400


@app.route('/stop', methods=['POST'])
def stop_music():
    data = request.get_json()
    guild_id = int(data['guild_id'])

    async def _stop():
        voice_client = discord.utils.get(bot.voice_clients, guild_id=guild_id)
        if voice_client:
            # キューをクリアして切断
            if guild_id in queues:
                queues[guild_id] = []
            await voice_client.disconnect()

    future = asyncio.run_coroutine_threadsafe(_stop(), bot.loop)
    future.result()
    return jsonify({'status': 'stopped and disconnected'})

# --- サーバーの起動 ---
def run_flask():
    app.run(host='0.0.0.0', port=5002)

def run_bot():
    # .envからTOKENを読み込む
    from dotenv import load_dotenv
    import os
    load_dotenv()
    bot_token = os.getenv("DISCORD_BOT_TOKEN")
    if bot_token:
        bot.run(bot_token)
    else:
        print("Error: DISCORD_BOT_TOKEN not found in .env file")

if __name__ == '__main__':
    # Flaskサーバーを別スレッドで起動
    flask_thread = threading.Thread(target=run_flask)
    flask_thread.daemon = True
    flask_thread.start()

    # Discord Botをメインスレッドで起動
    run_bot()